File: bundler/src/com/hendraanggrian/bundler/internal/BundleBinding.java
package com.hendraanggrian.bundler.internal;

import android.os.Bundle;
import android.os.Parcelable;
import android.util.SparseArray;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.Callable;

import androidx.annotation.RequiresApi;

public abstract class BundleBinding {

    public static final BundleBinding EMPTY = new BundleBinding(Bundle.EMPTY) {
    };

    protected Bundle source;
    private List<?> args;

    protected BundleBinding(Bundle source) {
        this(source, Collections.emptyList());
    }

    protected BundleBinding(List<?> args) {
        this(new Bundle(), args);
    }

    private BundleBinding(Bundle source, List<?> args) {
        this.source = source;
        this.args = args;
    }

    public Bundle getSource() {
        return source;
    }

    // region Non-void primitive types: supports unboxed, boxed, and unboxed array (only int supports ArrayList).
    protected boolean getBoolean(String key, boolean defaultValue) {
        return source.getBoolean(key, defaultValue);
    }

    protected Boolean getBoolean(final String key, Boolean defaultValue) {
        return find(key, defaultValue, new Callable<Boolean>() {
            @Override
            public Boolean call() {
                return source.getBoolean(key);
            }
        });
    }

    protected boolean[] getBooleanArray(final String key, boolean[] defaultValue) {
        return find(key, defaultValue, new Callable<boolean[]>() {
            @Override
            public boolean[] call() {
                return source.getBooleanArray(key);
            }
        });
    }

    protected byte getByte(String key, byte defaultValue) {
        return source.getByte(key, defaultValue);
    }

    protected Byte getByte(final String key, Byte defaultValue) {
        return find(key, defaultValue, new Callable<Byte>() {
            @Override
            public Byte call() {
                return source.getByte(key);
            }
        });
    }

    protected byte[] getByteArray(final String key, byte[] defaultValue) {
        return find(key, defaultValue, new Callable<byte[]>() {
            @Override
            public byte[] call() {
                return source.getByteArray(key);
            }
        });
    }

    protected char getChar(String key, char defaultValue) {
        return source.getChar(key, defaultValue);
    }

    protected Character getChar(final String key, Character defaultValue) {
        return find(key, defaultValue, new Callable<Character>() {
            @Override
            public Character call() {
                return source.getChar(key);
            }
        });
    }

    protected char[] getCharArray(final String key, char[] defaultValue) {
        return find(key, defaultValue, new Callable<char[]>() {
            @Override
            public char[] call() {
                return source.getCharArray(key);
            }
        });
    }

    protected double getDouble(String key, double defaultValue) {
        return source.getDouble(key, defaultValue);
    }

    protected Double getDouble(final String key, Double defaultValue) {
        return find(key, defaultValue, new Callable<Double>() {
            @Override
            public Double call() {
                return source.getDouble(key);
            }
        });
    }

    protected double[] getDoubleArray(final String key, double[] defaultValue) {
        return find(key, defaultValue, new Callable<double[]>() {
            @Override
            public double[] call() {
                return source.getDoubleArray(key);
            }
        });
    }

    protected float getFloat(String key, float defaultValue) {
        return source.getFloat(key, defaultValue);
    }

    protected Float getFloat(final String key, Float defaultValue) {
        return find(key, defaultValue, new Callable<Float>() {
            @Override
            public Float call() {
                return source.getFloat(key);
            }
        });
    }

    protected float[] getFloatArray(final String key, float[] defaultValue) {
        return find(key, defaultValue, new Callable<float[]>() {
            @Override
            public float[] call() {
                return source.getFloatArray(key);
            }
        });
    }

    protected int getInt(String key, int defaultValue) {
        return source.getInt(key, defaultValue);
    }

    protected Integer getInt(final String key, Integer defaultValue) {
        return find(key, defaultValue, new Callable<Integer>() {
            @Override
            public Integer call() {
                return source.getInt(key);
            }
        });
    }

    protected int[] getIntArray(final String key, int[] defaultValue) {
        return find(key, defaultValue, new Callable<int[]>() {
            @Override
            public int[] call() {
                return source.getIntArray(key);
            }
        });
    }

    protected ArrayList<Integer> getIntegerArrayList(final String key, ArrayList<Integer> defaultValue) {
        return find(key, defaultValue, new Callable<ArrayList<Integer>>() {
            @Override
            public ArrayList<Integer> call() {
                return source.getIntegerArrayList(key);
            }
        });
    }

    protected long getLong(String key, long defaultValue) {
        return source.getLong(key, defaultValue);
    }

    protected Long getLong(final String key, Long defaultValue) {
        return find(key, defaultValue, new Callable<Long>() {
            @Override
            public Long call() {
                return source.getLong(key);
            }
        });
    }

    protected long[] getLongArray(final String key, long[] defaultValue) {
        return find(key, defaultValue, new Callable<long[]>() {
            @Override
            public long[] call() {
                return source.getLongArray(key);
            }
        });
    }

    protected short getShort(String key, short defaultValue) {
        return source.getShort(key, defaultValue);
    }

    protected Short getShort(final String key, Short defaultValue) {
        return find(key, defaultValue, new Callable<Short>() {
            @Override
            public Short call() {
                return source.getShort(key);
            }
        });
    }

    protected short[] getShortArray(final String key, short[] defaultValue) {
        return find(key, defaultValue, new Callable<short[]>() {
            @Override
            public short[] call() {
                return source.getShortArray(key);
            }
        });
    }
    // endregion

    // region Non-primitive types: supports single object, array, and ArrayList (only Parcelable supports SparseArray).
    protected CharSequence getCharSequence(String key, CharSequence defaultValue) {
        return source.getCharSequence(key, defaultValue);
    }

    protected CharSequence[] getCharSequenceArray(final String key, CharSequence[] defaultValue) {
        return find(key, defaultValue, new Callable<CharSequence[]>() {
            @Override
            public CharSequence[] call() {
                return source.getCharSequenceArray(key);
            }
        });
    }

    protected ArrayList<CharSequence> getCharSequenceArrayList(final String key, ArrayList<CharSequence> defaultValue) {
        return find(key, defaultValue, new Callable<ArrayList<CharSequence>>() {
            @Override
            public ArrayList<CharSequence> call() {
                return source.getCharSequenceArrayList(key);
            }
        });
    }

    protected <T extends Parcelable> T getParcelable(final String key, T defaultValue) {
        return find(key, defaultValue, new Callable<T>() {
            @Override
            public T call() {
                return source.getParcelable(key);
            }
        });
    }

    protected <T extends Parcelable> T[] getParcelableArray(final String key, T[] defaultValue) {
        return find(key, defaultValue, new Callable<T[]>() {
            @Override
            @SuppressWarnings("unchecked")
            public T[] call() {
                return (T[]) source.getParcelableArray(key);
            }
        });
    }

    protected <T extends Parcelable> ArrayList<T> getParcelableArrayList(final String key, ArrayList<T> defaultValue) {
        return find(key, defaultValue, new Callable<ArrayList<T>>() {
            @Override
            public ArrayList<T> call() {
                return source.getParcelableArrayList(key);
            }
        });
    }

    protected <T extends Parcelable> SparseArray<T> getSparseParcelableArray(final String key, SparseArray<T> defaultValue) {
        return find(key, defaultValue, new Callable<SparseArray<T>>() {
            @Override
            public SparseArray<T> call() {
                return source.getSparseParcelableArray(key);
            }
        });
    }

    @RequiresApi(12)
    protected String getString(String key, String defaultValue) {
        return source.getString(key, defaultValue);
    }

    protected String[] getStringArray(final String key, String[] defaultValue) {
        return find(key, defaultValue, new Callable<String[]>() {
            @Override
            public String[] call() {
                return source.getStringArray(key);
            }
        });
    }

    protected ArrayList<String> getStringArrayList(final String key, ArrayList<String> defaultValue) {
        return find(key, defaultValue, new Callable<ArrayList<String>>() {
            @Override
            public ArrayList<String> call() {
                return source.getStringArrayList(key);
            }
        });
    }
    // endregion

    // region Others: Parceler and Serializable.
    protected <T extends Serializable> T getSerializable(final String key, T defaultValue) {
        return find(key, defaultValue, new Callable<T>() {
            @Override
            @SuppressWarnings("unchecked")
            public T call() {
                return (T) source.getSerializable(key);
            }
        });
    }
    // endregion

    protected void checkRequired(String key, String targetName) {
        if (!source.containsKey(key)) {
            throw new IllegalStateException(String.format(
                "Required extra %s with key %s not found, " +
                    "if this extra is optional add @Nullable to this field.",
                targetName, key));
        }
    }

    protected Object nextArg() {
        final Object arg = args.get(0);
        args.remove(0);
        return arg;
    }

    private <T> T find(String key, T defaultValue, Callable<T> callable) {
        if (source.containsKey(key)) {
            try {
                return callable.call();
            } catch (Exception e) {
                throw new IllegalStateException();
            }
        }
        return defaultValue;
    }
}

File: bundler/src/com/hendraanggrian/bundler/internal/BundleBinding.kt
package com.hendraanggrian.bundler.internal

import android.os.Bundle
import android.os.Parcelable
import android.util.SparseArray
import java.io.Serializable

abstract class BundleBinding private constructor(val source: Bundle, private val args: MutableList<*>) {

    companion object {
        val EMPTY: BundleBinding = object : BundleBinding(Bundle.EMPTY) {}
    }

    protected constructor(source: Bundle) : this(source, mutableListOf<Any>())
    protected constructor(args: MutableList<*>) : this(Bundle(), args)

    protected fun getBoolean(key: String, defaultValue: Boolean): Boolean = source.getBoolean(key, defaultValue)
    protected fun getBooleanArray(key: String, defaultValue: BooleanArray): BooleanArray =
        findValue(key, defaultValue) { source.getBooleanArray(key) }

    protected fun getByte(key: String, defaultValue: Byte): Byte = source.getByte(key, defaultValue)
    protected fun getByteArray(key: String, defaultValue: ByteArray): ByteArray =
        findValue(key, defaultValue) { source.getByteArray(key) }

    protected fun getChar(key: String, defaultValue: Char): Char = source.getChar(key, defaultValue)
    protected fun getCharArray(key: String, defaultValue: CharArray): CharArray =
        findValue(key, defaultValue) { source.getCharArray(key) }

    protected fun getDouble(key: String, defaultValue: Double): Double = source.getDouble(key, defaultValue)
    protected fun getDoubleArray(key: String, defaultValue: DoubleArray): DoubleArray =
        findValue(key, defaultValue) { source.getDoubleArray(key) }

    protected fun getFloat(key: String, defaultValue: Float): Float = source.getFloat(key, defaultValue)
    protected fun getFloatArray(key: String, defaultValue: FloatArray): FloatArray =
        findValue(key, defaultValue) { source.getFloatArray(key) }

    protected fun getInt(key: String, defaultValue: Int): Int = source.getInt(key, defaultValue)
    protected fun getIntArray(key: String, defaultValue: IntArray): IntArray =
        findValue(key, defaultValue) { source.getIntArray(key) }

    protected fun getLong(key: String, defaultValue: Long): Long = source.getLong(key, defaultValue)
    protected fun getLongArray(key: String, defaultValue: LongArray): LongArray =
        findValue(key, defaultValue) { source.getLongArray(key) }

    protected fun getShort(key: String, defaultValue: Short): Short = source.getShort(key, defaultValue)
    protected fun getShortArray(key: String, defaultValue: ShortArray): ShortArray =
        findValue(key, defaultValue) { source.getShortArray(key) }

    protected fun getCharSequence(key: String, defaultValue: CharSequence): CharSequence =
        source.getCharSequence(key, defaultValue)

    protected fun getCharSequenceArray(key: String, defaultValue: Array<CharSequence>): Array<CharSequence> =
        findValue(key, defaultValue) { source.getCharSequenceArray(key) }

    protected fun getCharSequenceArrayList(
        key: String,
        defaultValue: ArrayList<CharSequence>
    ): ArrayList<CharSequence> = findValue(key, defaultValue) { source.getCharSequenceArrayList(key) }

    protected fun <T : Parcelable> getParcelable(key: String, defaultValue: T): T =
        findValue(key, defaultValue) { source.getParcelable(key) }

    protected fun <T : Parcelable> getParcelableArray(key: String, defaultValue: Array<T>): Array<T> =
        findValue(key, defaultValue) { source.getParcelableArray(key) as Array<T> }

    protected fun <T : Parcelable> getParcelableArrayList(key: String, defaultValue: ArrayList<T>): ArrayList<T> =
        findValue(key, defaultValue) { source.getParcelableArrayList(key) }

    protected fun <T : Parcelable> getSparseParcelableArray(key: String, defaultValue: SparseArray<T>): SparseArray<T> =
        findValue(key, defaultValue) { source.getSparseParcelableArray(key) }

    protected fun getString(key: String, defaultValue: String): String = source.getString(key, defaultValue)
    protected fun getStringArray(key: String, defaultValue: Array<String>): Array<String> =
        findValue(key, defaultValue) { source.getStringArray(key) }

    protected fun getStringArrayList(key: String, defaultValue: ArrayList<String>): ArrayList<String> =
        findValue(key, defaultValue) { source.getStringArrayList(key) }

    protected fun <T : Serializable> getSerializable(key: String, defaultValue: T): T =
        findValue(key, defaultValue) { source.getSerializable(key) as T }

    protected fun checkRequired(key: String, targetName: String) {
        check(source.containsKey(key)) {
            error(
                "Required extra `$targetName` with key `$key` not found, " +
                    "if this extra is optional add @Nullable to this field."
            )
        }
    }

    protected fun nextArg(): Any? {
        val arg = args[0]
        args.removeAt(0)
        return arg
    }

    private inline fun <T> findValue(key: String, defaultValue: T, getValue: (key: String) -> T?): T {
        if (source.containsKey(key)) {
            val value = getValue(key)
            if (value != null) {
                return value
            }
        }
        return defaultValue
    }
}
